provision/kubernetes/logs.go

```diff
@@ -0,0 +1,318 @@
+		go func(index int, pod *apiv1.Pod) {
+			defer wg.Done()
+
+			request := clusterClient.CoreV1().Pods(ns).GetLogs(pod.ObjectMeta.Name, &apiv1.PodLogOptions{
+				TailLines:  tailLimit,
+				Timestamps: true,
+			})
+			stream, err := request.Stream()
+			if err != nil {
+				errs[index] = err
+				return
+			}
+
+			appName := pod.ObjectMeta.Labels["tsuru.io/app-name"]
+			appProcess := pod.ObjectMeta.Labels["tsuru.io/app-process"]
+
+			scanner := bufio.NewScanner(stream)
+			tsuruLogs := make([]appTypes.Applog, 0)
+			for scanner.Scan() {
+				line := scanner.Text()
+
+				if len(line) == 0 {
+					continue
+				}
+				tsuruLog := parsek8sLogLine(line)
+				tsuruLog.Unit = pod.ObjectMeta.Name
+				tsuruLog.AppName = appName
+				tsuruLog.Source = appProcess
+				tsuruLogs = append(tsuruLogs, tsuruLog)
+			}
+
+			if err := scanner.Err(); err != nil && !knet.IsProbableEOF(err) {
+				errs[index] = err
+			}
+
+			logs[index] = tsuruLogs
+		}(index, pod)
+	}
+
+	wg.Wait()
+
+	unifiedLog := []appTypes.Applog{}
+	for _, podLogs := range logs {
+		unifiedLog = append(unifiedLog, podLogs...)
+	}
+
+	sort.Slice(unifiedLog, func(i, j int) bool { return unifiedLog[i].Date.Before(unifiedLog[j].Date) })
+
+	for index, err := range errs {
+		if err == nil {
+			continue
+		}
+
+		pod := pods[index]
+		appName := pod.ObjectMeta.Labels["tsuru.io/app-name"]
+
+		unifiedLog = append(unifiedLog, errToLog(pod.ObjectMeta.Name, appName, err))
+	}
+
+	return unifiedLog, nil
+}
+
+func listPodsSelectorForLog(args appTypes.ListLogArgs) labels.Selector {
+	return labels.SelectorFromSet(labels.Set(map[string]string{
+		"tsuru.io/app-name": args.AppName,
+	}))
+}
+
+func parsek8sLogLine(line string) (appLog appTypes.Applog) {
+	parts := strings.SplitN(line, logLineTimeSeparator, 2)
+
+	if len(parts) < 2 {
+		appLog.Message = string(line)
+		return
+	}
+	appLog.Date, _ = parseRFC3339(parts[0])
+	appLog.Message = parts[1]
+
+	return
+}
+
+func parseRFC3339(s string) (time.Time, error) {
+	if t, timeErr := time.Parse(time.RFC3339Nano, s); timeErr == nil {
+		return t, nil
+	}
+	return time.Parse(time.RFC3339, s)
+}
+
+func tailLines(i int) *int64 {
+	b := int64(i)
+	return &b
+}
+
+func errToLog(podName, appName string, err error) appTypes.Applog {
+	return appTypes.Applog{
+		Date:    time.Now().UTC(),
+		Message: fmt.Sprintf("Could not get logs from unit: %s, error: %s", podName, err.Error()),
+		Unit:    "apiserver",
+		AppName: appName,
+		Source:  "kubernetes",
+	}
+}
+
+func infoToLog(appName string, message string) appTypes.Applog {
+	return appTypes.Applog{
+		Date:    time.Now().UTC(),
+		Message: message,
+		Unit:    "apiserver",
+		AppName: appName,
+		Source:  "kubernetes",
+	}
+}
+
+type k8sLogsNotifier struct {
+	ctx  context.Context
+	ch   chan appTypes.Applog
+	ns   string
+	wg   sync.WaitGroup
+	once sync.Once
+	done context.CancelFunc
+
+	clusterClient *ClusterClient
+}
+
+func (k *k8sLogsNotifier) watchPod(pod *apiv1.Pod, notify bool) {
+	k.wg.Add(1)
```

Comment:
This is dangerous and could cause unexpected behavior. Waitgroups have one weird quirk, according to the [docs](https://golang.org/pkg/sync/#WaitGroup.Add):

> Note that calls with a positive delta that occur when the counter is zero must happen before a Wait.

Since this is called in a goroutine there's a chance that this code will only be executed after the call to Wait(). This would cause the Wait() call to return immediately and assume everything has finished.

I think it would be safer to increment the waitgroup in the caller, before starting the goroutine.